File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
import java.util.*;
import java.lang.*;
class Solution {
    private static class State {
    public final String word;
    public final int step;
    private State(String word, int step) {
        this.word = word;
        this.step = step;
    }
    }
    public int solution(String begin, String target, String[] words) {
        boolean[] isVisited = new boolean[words.length];
        Queue<State> queue = new LinkedList<>();
        queue.add(new State(begin,0));
        
        while(!queue.isEmpty()) {
            State state = queue.poll();
            if(state.word.equals(target)) {
                return state.step;
            }
            for(int i = 0; i < words.length; i++){
                String next = words[i];
                 if(!isConvertable(state.word, next)) {
                    continue;
                 }
                if(isVisited[i]) continue;
                isVisited[i] = true;
                queue.add(new State(next, state.step + 1));
            }
           
        }
        return 0;
       
    }
    private boolean isConvertable(String src, String dist) {
        char[] srcArr = src.toCharArray();
        char[] distArr = dist.toCharArray();
        int diff = 0;
        for(int i = 0; i < srcArr.length; i++) {
            if(srcArr[i] != distArr[i]) diff++;
        }
        return diff == 1;
    }
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    // 1. 방문 여부를 체크해주는 리스트를 만든다
    const isVisited = [];
    // 3. 큐를 만들어준다.
    const queue = [new State(begin,0)];
    // 4. 큐가 빌때까지 while을 돌린다.
    while(queue.length != 0) {
        const state = queue.shift();
        if(state.word === target) {
            return state.step;
        }
        for(let i = 0; i < words.length; i++){
            const next = words[i];
            if(!isConvertable(state.word, next)) {
                continue;
            }
            if(isVisited[i]) {
                continue;
            }
            isVisited[i] = true;
            queue.push(new State(next, state.step + 1));
        }
    }
    return 0;
}
// 2. 문자를 비교해주는 함수를 만든다.
function isConvertable(src, dist) {
    let diff = 0;
    for(let i = 0; i < src.length; i++) {
        if(src[i] != dist[i]) diff++;
    }
    return diff === 1;
}

class State {
    constructor(word, step) {
        this.word = word;
        this.step = step;
    }
}